executor/adapter.go

```diff
@@ -1492,6 +1502,92 @@ func resetCTEStorageMap(se sessionctx.Context) error {
+// Currently, only support "Selection + TableScan" pattern
+func (a *ExecStmt) RecordHistoryStats() {
+	sessVars := a.Ctx.GetSessionVars()
+	if !sessVars.EnableOptimizerHistoryStats || sessVars.InRestrictedSQL || !sessVars.StmtCtx.InSelectStmt {
+		return
+	}
+
+	sql := FormatSQL(a.GetTextToLog(true))()
+	stmtCtx := sessVars.StmtCtx
+	flat := getFlatPlan(stmtCtx)
+	if flat == nil || len(flat.Main) == 0 {
+		return
+	}
+
+	for _, flatCTEPlanTree := range flat.CTEs {
+		a.traverseFlatPlanForHistoryStats(stmtCtx, flatCTEPlanTree, sql)
+	}
+	for _, flatSubQueryPlanTree := range flat.ScalarSubQueries {
+		a.traverseFlatPlanForHistoryStats(stmtCtx, flatSubQueryPlanTree, sql)
+	}
+	a.traverseFlatPlanForHistoryStats(stmtCtx, flat.Main, sql)
+}
+
+func (a *ExecStmt) traverseFlatPlanForHistoryStats(stmtCtx *stmtctx.StatementContext, flatPlanTree plannercore.FlatPlanTree, sql string) {
+	for _, flatOp := range flatPlanTree {
+		pp, isPhysicalPlan := flatOp.Origin.(plannercore.PhysicalPlan)
+		if !isPhysicalPlan {
+			continue
+		}
+		switch x := pp.(type) {
+		case *plannercore.PhysicalSelection:
+			if tableScan, ok := x.Children()[0].(*plannercore.PhysicalTableScan); ok {
+				if tableScan.HasRFFilters() {
+					break
+				}
+				needRecord, estRows, actualRows := needRecordHistoryStats(stmtCtx, x)
+				if !needRecord {
+					break
+				}
+				succ, hashCode := genHistoryStatsHashCodeForSelection(stmtCtx, x)
+				if !succ {
+					break
+				}
+				recordHistoryStats(sql, tableScan.Table.ID, x.ID(), false, hashCode, estRows, actualRows)
+			}
+		case *plannercore.PhysicalTableScan:
+			if x.HasRFFilters() {
+				break
+			}
+			needRecord, estRows, actualRows := needRecordHistoryStats(stmtCtx, x)
+			if !needRecord {
+				break
+			}
+			recordHistoryStats(sql, x.Table.ID, x.ID(), true, 0, estRows, actualRows)
+		}
+	}
+}
+
+func recordHistoryStats(sql string, tableID int64, planID int, nullHashCode bool, hashCode uint64, estRows float64, actualRows int64) {
+	logutil.BgLogger().Debug(
+		sql,
+		zap.Int64("TableID", tableID),
+		zap.Int("PlanID", planID),
+		zap.Bool("NullHashCode", nullHashCode),
+		zap.Uint64("FinalHash", hashCode),
+		zap.Float64("EstRows", estRows),
+		zap.Int64("ActRows", actualRows))
+}
+
+func genHistoryStatsHashCodeForSelection(stmtCtx *stmtctx.StatementContext, x *plannercore.PhysicalSelection) (bool, uint64) {
+	return expression.SortAndGenMD5HashForCNFExprs(stmtCtx, x.Conditions)
+}
+
+func needRecordHistoryStats(stmtCtx *stmtctx.StatementContext, x plannercore.PhysicalPlan) (bool, float64, int64) {
+	estRows := plannercore.GetPhysicalPlanEstRowCount(x)
+	hasCopStats, actualRows := GetResultRowsCountForCopStats(stmtCtx, x.ID())
+	if !hasCopStats {
+		return false, estRows, actualRows
+	}
+	actualRowsFloat := float64(actualRows)
+	max := math.Max(actualRowsFloat, estRows)
+	min := math.Min(actualRowsFloat, estRows)
+	return max > min*HistoryStatsUpdateThreshold, estRows, actualRows
```

Comment:
What if the actRows is 2 and estRows is 1, do we still need to update the history stats?

Reply:
Good question, according to Ahmad, even if row count is very small, he believes the ratio value is meaningful for optimization.